ffi/src/x509.rs

```diff
@@ -215,4 +215,188 @@ pub mod ffi {
             Ok(Box::new(Cert(cert)))
         }
     }
+
+    #[diplomat::opaque]
+    pub struct CertVec(pub Vec<certificate::Cert>);
+
+    impl CertVec {
+        pub fn next(&mut self) -> Option<Box<Cert>> {
+            self.0.pop().map(|cert| Box::new(Cert(cert)))
+        }
+    }
+
+    #[diplomat::opaque]
+    pub struct EncapsulatedContentInfo(pub picky_asn1_x509::content_info::EncapsulatedContentInfo);
+
+    impl EncapsulatedContentInfo {
+        pub fn content_type(&self, writable: &mut DiplomatWriteable) -> Result<(), Box<PickyError>> {
+            let oid: String = self.0.content_type.0.clone().into();
+            write!(writable, "{}", oid)?;
+            Ok(())
+        }
+    }
+
+    /// TODO: expose the SignerInfo struct public fields
+    #[diplomat::opaque]
+    pub struct SignerInfo(pub picky_asn1_x509::signer_info::SignerInfo);
+
+    #[diplomat::opaque]
+    pub struct SignerInfos(pub Vec<SignerInfo>);
+
+    impl SignerInfos {
+        pub fn next(&mut self) -> Option<Box<SignerInfo>> {
+            self.0.pop().map(Box::new)
+        }
+    }
+
+    #[diplomat::opaque]
+    pub struct Der(pub Vec<u8>);
+
+    impl Der {
+        pub fn is_empty(&self) -> bool {
+            self.0.is_empty()
+        }
+
+        pub fn len(&self) -> usize {
+            self.0.len()
+        }
+
+        pub fn fill<'a>(&'a self, buf: &'a mut [u8]) -> Result<(), Box<PickyError>> {
+            if buf.len() < self.0.len() {
+                return Err("Buffer too small".into());
+            }
+            buf.copy_from_slice(&self.0);
+            Ok(())
+        }
+    }
+
+    #[diplomat::opaque]
+    pub struct AlgorithmIdentifier(pub picky::AlgorithmIdentifier);
+
+    impl AlgorithmIdentifier {
+        pub fn is_a(&self, other: &str) -> Result<bool, Box<PickyError>> {
+            Ok(self
+                .0
+                .is_a(picky::oid::ObjectIdentifier::try_from(other).map_err(|_| "invalid OID")?))
+        }
+
+        pub fn oid(&self, writable: &mut DiplomatWriteable) -> Result<(), Box<PickyError>> {
```

Comment:
naming: unlike usual Rust convention, you should use the `get` prefix (`get_oid`). This will generator the C# properties too and we get more idiomatic C# code. Same for the other getters.

Reply:
updated